src/mlpack/core/tree/cosine_tree/cosine_tree.cpp

```diff
@@ -150,6 +150,197 @@ CosineTree::CosineTree(const arma::mat& dataset,
   ConstructBasis(treeQueue);
 }
 
+//! Copy the given tree.
+CosineTree::CosineTree(const CosineTree& other) :
+    dataset(other.parent->GetDataset()),
+    delta(other.delta),
+    parent(other.Parent()),
+    left(other.Left()),
+    right(other.Right()),
+    indices(other.indices),
+    l2NormsSquared(other.l2NormsSquared),
+    centroid(other.centroid),
+    basisVector(other.basisVector),
+    splitPointIndex(other.SplitPointIndex()),
+    numColumns(other.NumColumns()),
+    l2Error(other.L2Error()),
+    frobNormSquared(other.FrobNormSquared())
+{
+  // Create left and right children (if any).
+  if (other.Left())
+  {
+    left = new CosineTree(*other.Left());
+    left->Parent() = this; // Set parent to this, not other tree.
+  }
+
+  if (other.Right())
+  {
+    right = new CosineTree(*other.Right());
+    right->Parent() = this; // Set parent to this, not other tree.
+  }
+
+  // Propagate matrix, but only if we are the root.
+  if (parent == NULL)
+  {
+    std::queue<CosineTree*> queue;
+    if (left)
+      queue.push(left);
+    if (right)
+      queue.push(right);
+    while (!queue.empty())
+    {
+      CosineTree* node = queue.front();
+      queue.pop();
+
+      node->dataset = dataset;
+      if (node->left)
+        queue.push(node->left);
+      if (node->right)
+        queue.push(node->right);
+    }
+  }
+}
+
+//! Copy Assignment.
+CosineTree&CosineTree::operator=(const CosineTree& other)
+{
+  // Return if it's the same tree.
+  if (this == &other)
+    return *this;
+
+  // Freeing memory that will not be used anymore.  
+  delete left;
+  delete right;
+
+  dataset = (other.parent == NULL) ? other.parent->GetDataset() : NULL;
+  delta = other.delta;
+  parent = other.Parent();
+  left = other.Left();
+  right = other.Right();
+  indices = other.indices;
+  l2NormsSquared = other.l2NormsSquared;
+  centroid = other.centroid;
+  basisVector = other.basisVector;
+  splitPointIndex = other.SplitPointIndex();
+  numColumns = other.NumColumns();
+  l2Error = other.L2Error();
+  frobNormSquared = other.FrobNormSquared();
+
+  // Create left and right children (if any).
+  if (other.Left())
+  {
+    left = new CosineTree(*other.Left());
+    left->Parent() = this; // Set parent to this, not other tree.
+  }
+
+  if (other.Right())
+  {
+    right = new CosineTree(*other.Right());
+    right->Parent() = this; // Set parent to this, not other tree.
+  }
+
+  // Propagate matrix, but only if we are the root.
+  if (parent == NULL)
+  {
+    std::queue<CosineTree*> queue;
+    if (left)
+      queue.push(left);
+    if (right)
+      queue.push(right);
+    while (!queue.empty())
+    {
+      CosineTree* node = queue.front();
+      queue.pop();
+
+      node->dataset = dataset;
+      if (node->left)
+        queue.push(node->left);
+      if (node->right)
+        queue.push(node->right);
+    }
+  }
+
+  return *this;
+}
+
+//! Move the given tree.
+CosineTree::CosineTree(CosineTree&& other) :
+  dataset(other.dataset),
```

Comment:
```suggestion
    dataset(other.dataset),
```

Can you make these lines indented doubly?  They are wrapped from the previous line, so that's what the style guide suggests. :+1: